### PR TITLE
Use globals for the storage of the kernel's pagetable, physical memory manager and console

### PR DESCRIPTION
**Pull request by n1tram1**
probably has lot of useless code, also has debug prints by the bucketload

### CODE REVIEW

File: kernel/src/mm/physical_memory_manager.rs [hunk 1]
```diff
@@ -304,6 +317,32 @@ impl PhysicalMemoryManager {
...
+        if unsafe { globals::STATE.is_mmu_enabled() } {
+            kprintln!("alloc_rw_pages: #{} pages -> {:X}", page_count, addr);
+            globals::KERNEL_PAGETABLE.lock(|kernel_pt| {
+                for page_addr in (addr..(addr + page_count * self.page_size())).step_by(self.page_size()) {
+                    kernel_pt
+                        .map(page_addr.into(), page_addr.into(), mm::Permissions::READ | mm::Permissions::WRITE)
+                        .unwrap();
+                    }
+            });
+            kprintln!("\talloc_rw_pages: returning {:X}", addr);
```
Comment (Skallwar) [inline kernel/src/mm/physical_memory_manager.rs:338]:
```suggestion
```

File: kernel/src/mm/physical_memory_manager.rs [hunk 2]
```diff
@@ -304,6 +317,32 @@ impl PhysicalMemoryManager {
...
+        // If there is a kernel pagetable, identity map the pages.
+        // Not sure this is the best idea, but I would say it follows the
+        // principle of least astonishment.
+        //
+        let first_page = self.alloc_pages(page_count)?;
+        let addr: usize = first_page.into();
+
+
+        if unsafe { globals::STATE.is_mmu_enabled() } {
+            kprintln!("alloc_rw_pages: #{} pages -> {:X}", page_count, addr);
```
Comment (Skallwar) [inline kernel/src/mm/physical_memory_manager.rs:330]:
```suggestion
```

File: Cargo.toml [hunk 3]
```diff
@@ -5,4 +5,5 @@ members = [
     "drivers",
     "aarch64_qemuvirt",
     "riscv64_qemuvirt",
+    "freetos",
```
Comment (Skallwar) [inline Cargo.toml:8]:
```suggestion
```

**Review by Skallwar — COMMENTED** [review]
`core_local.rs` is empty, let's remove it

File: drivers/src/null_uart.rs [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+
+impl NullUart {
+    pub const fn new() -> Self {
+        Self
+    }
+}
+
+impl Console for NullUart {
+    fn write(&self, _data: &str) {
+        // Does nothing, just a placeholder while a real uart is not in place.
```
[the remark this answers is not in the document]
Comment (n1tram1) [reply drivers/src/null_uart.rs:14]: Wouldn't work. The idea is this is a dummy console while the proper console driver hasn't been inserted. If something tries to print, we wouldn't want to panic the kernel.
In an ideal world we would buffer output and then print it once we have proper console.

File: kernel/src/arch/aarch64/pgt48.rs [hunk 5]
```diff
@@ -198,25 +200,48 @@ union PageTableContent {
...
 }
 
 impl PageTable {
+    pub const fn zeroed() -> Self {
+        let mut entries: [PageTableContent; 512] = unsafe { core::mem::MaybeUninit::uninit().assume_init() };
+        let mut i = 0;
+        while i < 512 {
+            entries[i] = PageTableContent::new_invalid();
+            i += 1;
+        }
```
[the remark this answers is not in the document]
Comment (n1tram1) [reply kernel/src/arch/aarch64/pgt48.rs:222]: I love this 💜
But it is not const fn

File: kernel/src/globals.rs [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+use crate::lock::Lock;
+
+use crate::mm;
+
+use drivers::init_cell::InitCell;
+use drivers::Console;
+
+static NULL_CONSOLE: drivers::null_uart::NullUart = drivers::null_uart::NullUart::new();
+
+pub static CONSOLE: InitCell<&'static (dyn drivers::Console + Sync)> = InitCell::new(&NULL_CONSOLE);
```
[the remark this answers is not in the document]
Comment (n1tram1) [reply kernel/src/globals.rs:12]: Not needed, the type signature forces the object to be sync, so it is already thread safe.

File: kernel/src/lib.rs [hunk 7]
```diff
@@ -1,12 +1,20 @@
 #![no_std]
 #![feature(naked_functions)]
 #![feature(fn_align)]
+#![feature(const_mut_refs)]
+#![feature(slice_ptr_get)]
+#![feature(const_ptr_as_ref)]
+#![feature(const_slice_from_raw_parts_mut)]
+#![feature(iterator_try_collect)]
+#![feature(const_for)]
```
[the remark this answers is not in the document]
Comment (n1tram1) [reply kernel/src/lib.rs:9]: I think at least half of those could be removed 😅

**Comment by n1tram1** [comment]
Moved to https://github.com/ks0n/goOSe/pull/122

**Pull request closed without merging**